src/library/scala/collection/mutable/Builder.scala

```diff
@@ -164,7 +177,15 @@ class StringBuilder(private val sb: java.lang.StringBuilder) extends AbstractSeq
     *  @param  xs  the characters to be appended.
     *  @return     this StringBuilder.
     */
-  def appendAll(xs: IterableOnce[Char]): StringBuilder = appendAll(ArrayBuffer.from(xs).toArray)
+  def appendAll(xs: IterableOnce[Char]): StringBuilder = {
```

Comment:
Both implementations forego any optimizations that could be gained from `xs.copyToArray` as done in `ArrayBuilder`. We should implement them in `ArrayBuffer.from`.

This method here could also benefit from them to a lesser extent. You'd have to allocate an intermediate array though. Maybe that was the original idea behind using `ArrayBufer.from(xs).toArray` but it doesn't work because `ArrayBuffer` is always backed by an `Array[AnyRef]`. If a collection is backed by an `Array[Char]` I assume it will be much faster to use an intermediate array than to iterate over all boxed chars. The tricky part is knowing when to use which. Maybe using an array is still fast enough in other cases? (You still benefit from more efficient internal iteration when copying to the array)

Reply:
Are you suggesting using the ArrayBuilder code for all cases?

I'm having trouble following why copyToArray avoids the unboxing.

Reply:
If you have an `IterableOnce`, unless you can pattern match out that it is actually a wrapped array of char itself, your best bet is to simply walk the iterator manually, stuffing things into the `StringBuilder` as you go.  `StringBuilder` is at least as efficient per element as `ArrayBuffer`, and if you do it directly there are no extra steps.

Reply:
I'll add an appendAll(ArraySeq) method.  So the IterableOnce appendAll will only occur if scala can't determine the object type at compile. Do we think this will come up enough when using Arrays that it's worth doing runtime type checks?

Reply:
Is there an `appendAll(xs: Array[Char])`?  If not, there should be.  That's the most important case for speed.  Wrapped `ArraySeq[Char]` is fine too, but we don't yet know how heavily it will be used.

Reply:
https://docs.oracle.com/javase/8/docs/api/java/lang/StringBuilder.html#append-char:A-

Reply:
yes, scala StringBuilder has appendAll(<Array[Char]>)